### PR TITLE
Ticket3939 interpolation accuracy

### PR DESCRIPTION
**Pull request by LiamPanchaud**
New functionality to allow cvt record to convert using user defined method (used for spline interpolation in Keithley 2700)

### CODE REVIEW

File: cvtRecordApp/Makefile [hunk 1]
```diff
@@ -5,6 +5,8 @@ include $(TOP)/configure/CONFIG_APP
 
 DBDINC = cvtRecord menuCvtMethod menuCvtInitState
 
+USR_CFLAGS += /Od
+
```
Comment (FreddieAkeroyd) [inline cvtRecordApp/Makefile]: Is turning off optimisation required for it to work?

File: csmApp/Makefile [hunk 2]
```diff
@@ -5,6 +5,8 @@ LIBRARY_IOC = csmbase
 csmbase_SRCS = csmbase.c
 csmbase_LIBS += $(EPICS_BASE_IOC_LIBS)
 
+USR_CFLAGS += /Od
+
```
Comment (FreddieAkeroyd) [inline csmApp/Makefile]: Is turning off optimisation required for it to work?

Comment (LiamPanchaud) [reply csmApp/Makefile]: Removed

File: cvtRecordApp/Makefile (hunk 1 above)
Comment (LiamPanchaud) [reply cvtRecordApp/Makefile]: Removed

File: csmApp/csmbase.c [hunk 3]
```diff
@@ -1354,6 +1354,52 @@ epicsShareExtern double csm_z(csm_function *func, double x, double y)
...
+  \param arr the array to put the values into
+  \return 0 on success, -1 on error
+*/
+
+/*@EX(1)*/
+epicsShareExtern int get_arr_values(csm_function *func, char axis, double *arr, int arr_len) 
+{
+  switch(axis) {
+    case 'x':
+      for (int i = 0; i<arr_len; i++) {
```
Comment (Alistair-McGann-Tessella) [inline csmApp/csmbase.c:1372]: Could this segfault if the passed `arr_len` is different to the actual lengths of the arrays? If so, can there be any checking for this?

File: cvtRecordApp/cvtRecord.c [hunk 4]
```diff
@@ -515,25 +518,72 @@ static long readInputs(struct cvtRecord *pcvt)
...
         }
+        case menuCvtMethodUser1DTableSub:
+        {
+          double *x_arr;
+          double *y_arr;
+          int no_of_elements;
+
+          csm_function *csub; // dummy csub
+          char temp[BDIR_SIZE+TDIR_SIZE+SPEC_SIZE+2];
+          sprintf(temp, "%s/%s/%s", bdir, tdir, spec);
```
Comment (Alistair-McGann-Tessella) [inline cvtRecordApp/cvtRecord.c:537]: Does this print statement fill out the dummy csub with the data held in `temp`? I think this is an important part of this procedure but it is quite hard to follow at the minute

File: cvtRecordApp/cvtRecord.c [hunk 5]
```diff
@@ -515,25 +518,72 @@ static long readInputs(struct cvtRecord *pcvt)
...
+
+          csub = csm_new_function();
+
+          // contained within csub is an array of coordinate structs which need
+          // to be pulled out and given to the spline initialiser
+          if (!csm_read_1d_table(temp, csub)) {
```
Comment (Alistair-McGann-Tessella) [inline cvtRecordApp/cvtRecord.c:543]: I am very uncertain about where `temp` has come from here. I was trying to trace back where the config files get read in

**Review by Alistair-McGann-Tessella — CHANGES_REQUESTED** [review]

File: csmApp/csmbase.c (hunk 3 above)
Comment (LiamPanchaud) [reply csmApp/csmbase.c:1372]: Is _should_ never be different because it's retrieved directly from the struct which stores the array and the array length

File: cvtRecordApp/cvtRecord.c (hunk 4 above)
Comment (LiamPanchaud) [reply cvtRecordApp/cvtRecord.c:537]: This writes a string into `temp` consisting of `bdir` (a path), `tdir` (another part of the path), and `spec` (the file containing the look up specification at the end of the path)

**Review by Alistair-McGann-Tessella — APPROVED** [review]
